src/librustc_mir/transform/check_consts/validation.rs

```diff
@@ -248,12 +248,15 @@ impl Validator<'mir, 'tcx> {
             return;
         }
 
-        // If an operation is supported in miri (and is not already controlled by a feature gate) it
-        // can be turned on with `-Zunleash-the-miri-inside-of-you`.
-        let is_unleashable = O::IS_SUPPORTED_IN_MIRI && O::feature_gate().is_none();
+        // If an operation is supported in miri it can be turned on with
+        // `-Zunleash-the-miri-inside-of-you`.
+        let is_unleashable = O::IS_SUPPORTED_IN_MIRI;
 
         if is_unleashable && self.tcx.sess.opts.debugging_opts.unleash_the_miri_inside_of_you {
-            self.tcx.sess.span_warn(span, "skipping const checks");
+            self.tcx.sess.span_warn(self.tcx.def_span(self.def_id), "skipping const checks");
```

Comment:
side note: what do you think about emitting the feature name that is being skipped if there is one?

Reply:
Sure... or maybe not even emit the warning in that case? Either there will be other errors or a panic at the end listing the missing gates.

Reply:
I think emitting them is good, as it shows where things are being skipped. I mean, I added them on purpose so we'd be able so see exactly where a test is skipping checks, in order to notice when the checks change in any way

Reply:
Otherwise I'm worried this will kill deduplication, so we'll have 5 annotations per const again.

Reply:
> in order to notice when the checks change in any way

The point of the flag is to not worry about the static checks, and focus on testing the underlying engine. Why would I care if the checks change? I explicitly *do not want* to care about that, that's why I use the flag.^^

Reply:
Well... each of these is a true skipped static check. Unless you think there's no value to them? I personally found it convenient and helpful. I mean I was actually thinking now that it could make sense to emit "skipping `op specific message` const check" so we really know what's being skipped.

Reply:
For the way I am using the flag, these warnings provide negative value (i.e., they get in the way): I don't care what I am skipping, I just want to skip *all the checks* (it's called "unleash" for a reason :D ) -- and then I still have to acknowledge every skipped thing when writing UI tests, and adding all these `//~ WARN` is quite cumbersome.

Reply:
so.... how about we just emit one message at the end saying `skipped N static checks`?

Reply:
Oh or even better: we collect all of the messages with their spans in a list, and report a single warning at the end and give that warning lots of labels to point the messages there. This way you only have a single warning to skip and I get all my messages in the `.stderr` file

Reply:
Where would we collect that list?

Reply:
Maybe in the session variable you created? You don't need a list of feature gates anymore in that case, as you can just make it a boolean if any feature gate was skipped.

Reply:
How that? Don't I need a list of feature gates *and spans* so I can report them all?

Reply:
I was thinking of just stuffing the full message in there, so we can also start reporting a more detailed message for gateless checks.

Basically the error for not having feature gates in a build-pass unleash test would not contain the gate list anymore, but there'd be a separate warning in every unleash test (irrelevant of whether it uses feature gates or not or whether it compiles), and that warning would point to all skip sites.

Reply:
> and that warning would point to all skip sites.

That still requires a list of spans though, not just a `bool`? (And at that point we might as well track the gates as well.)

Reply:
@oli-obk I also do not understand whether you want these warnings to appear only in case the program otherwise builds, or always? If always, then sure we could just emit them here but writing miri-unleash UI tests would still be quite tedious. If only for "pass" tests, then... well see above, then the state needs to be stored somewhere so we can wait with emitting the warning until we know that rustc will actually succeed otherwise.

Maybe the flag should just have two values: one that emits the warning (eagerly, no extra state needed), and one that emits *no* warnings but also definitely will not let the build pass (like `delay_span_bug`, but actually delaying long enough).

Reply:
I was trying to avoid making the flag configurable. I indeed want the warning to always occur, and don't see why writing unleash tests would be tedious from that. All you get is a big warning at the end, which I consider ok, similar to how every MIR dump contains a warning about the instability of the dump.

Reply:
But then wouldn't we have to add `//~ WARN` everywhere for every single span of the warning, to make ui tests happy?

I don't really mind the warnings to show up when I run the tests (it's noise that can hide useful information, but whatever). The annoying part is getting the ui test suite to be happy.

Reply:
Also I still don't understand how "one warning at the end" could possibly be implemented with just a `bool` in the global state. I think you'll have to show some code to demonstrate that.

Reply:
> But then wouldn't we have to add //~ WARN everywhere for every single span of the warning, to make ui tests happy?

No, just a single one, probably at the top of the file, as it has no span (or maybe we use the first label's span). Labels don't have any representation in `//~` syntax, they just show up in stderr.

> Also I still don't understand how "one warning with all the information emitted at the end" could possibly be implemented with just a bool in the global state. I think you'll have to show some code to demonstrate that.

I don't mean just a bool, I mean to have both a bool and a list, but you don't need the `HashSet` for dedup, you just have a single list (potentially sort it at the end), that you always use to emit the warning. The bool is just there to emit an error in case the test would have successfully compiled but never have unleashed a non-feature-gated thing.

Reply:
> No, just a single one, probably at the top of the file, as it has no span (or maybe we use the first label's span). Labels don't have any representation in //~ syntax, they just show up in stderr.

I guess I don't know what "labels" vs "spans" are...

Reply:
> The bool is just there to emit an error in case the test would have successfully compiled but never have unleashed a non-feature-gated thing.

I thought we agreed in your PR that this is not the right kind of error, and instead we should error when unleashing any feature gate (like my PR does now)?

Reply:
> I thought we agreed in your PR that this is not the right kind of error, and instead we should error when unleashing any feature gate (like my PR does now)?

Sorry, we've went back and forth too many times. You're right.

Reply:
> I guess I don't know what "labels" vs "spans" are...

https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/struct.DiagnosticBuilder.html#method.span_label allows you to add an arbitrary annotation to an arbitrary span. This is what gives us the `^^^^^ something here` annotations.

Reply:
Hm... I did that, but it doesn't print the labels.
```rust
            for &(span, feature_gate) in unleashed_features.iter() {
                if let Some(feature_gate) = feature_gate {
                    diag.span_label(span, format!("skipping check for `{}` feature", feature_gate));
                } else {
                    diag.span_label(span, "skipping check that does not even have a feature gate");
                }
            }
            diag.emit();
```
becomes just
```
warning: skipping const checks
```

Reply:
I tried passing a proper span to `struct_warn`, no luck.

Our diagnostics APIs are quite hard to use correctly, it seems. :/

Reply:
I went for `span_help` now, that seems to work. It's quite the hack though...